File: java/src/org/rpcf/tests/sftest/TransferContext.java
package org.rpcf.tests.sftest;

import org.rpcf.rpcbase.IRpcService;
import org.rpcf.rpcbase.JRetVal;
import org.rpcf.rpcbase.RC;
import org.rpcf.rpcbase.rpcbase;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;

public class TransferContext {
    HashMap<Long,TransFileContext> m_mapChanToCtx = new HashMap<>();
    IRpcService m_oInst;

    TransferContext(IRpcService oInst)
    {m_oInst = oInst;}

    synchronized int addContext(long hChannel, TransFileContext o)
    {
        if(hChannel == RC.INVALID_HANDLE)
            return -RC.EINVAL;
        TransFileContext old =
            m_mapChanToCtx.get(hChannel);
        if(old != null && !old.isDone())
            return -RC.EEXIST;
        m_mapChanToCtx.put(hChannel,o);
        return RC.STATUS_SUCCESS;
    }

    synchronized void removeContext(long hChannel)
    { m_mapChanToCtx.remove(hChannel); }

    synchronized void onTransferDone(long hChannel, int iRet)
    {
        TransFileContext o = m_mapChanToCtx.get(hChannel);
        if(o == null)
            return;
        o.cleanUp();
        removeContext(hChannel);
    }

    synchronized int readFileAndSend(long hChannel)
    {
        int ret = 0;
        TransFileContext o = m_mapChanToCtx.get(hChannel);
        if(o == null)
            return -RC.ENOENT;
        int iSizeLimit = 2 * 1024 * 1024;
        try {
            while (true) {
                int iSize = o.m_iBytesLeft;
                if (iSize > iSizeLimit)
                    iSize = iSizeLimit;
                else if (iSize == 0) {
                    onTransferDone(hChannel, 0 );
                    break;
                }
                ByteBuffer bbuf = ByteBuffer.allocate(iSize);
                o.m_oChannel.read(bbuf);
                bbuf.flip();
                byte[] bufToSend = bbuf.array();
                ret = m_oInst.writeStreamAsync(hChannel, bufToSend);
                if(RC.isPending(ret)) {
                    ret = RC.STATUS_PENDING;
                    break;
                }
                if(RC.ERROR(ret))
                {
                    onTransferDone(hChannel,ret);
                    break;
                }
                o.m_iBytesLeft -= bufToSend.length;
                o.m_iBytesSent += bufToSend.length;

            }
        }catch( IOException e){
            ret = -RC.EIO;
            onTransferDone(hChannel,ret);
        }
        return ret;
    }

    int sendToken(long hChannel, byte[]buf )
    {
        return m_oInst.writeStreamNoWait(hChannel, buf);
    }
    synchronized int writeFileAndRecv(long hChannel, byte[] buf)
    {
        int ret = 0;
        if(hChannel == -RC.INVALID_HANDLE)
            return -RC.EINVAL;
        int iSizeLimit = 2 * 1024 * 1024;
        TransFileContext o = m_mapChanToCtx.get(hChannel);
        if(o == null)
            return -RC.ENOENT;

        try {
            while (true) {
                if(o.m_iBytesLeft < buf.length) {
                    ret = -RC.ERANGE;
                    break;
                }
                if(buf.length > 0) {
                    o.m_oFile.write(buf);
                    o.m_iBytesLeft -= buf.length;
                    o.m_iBytesSent += buf.length;

                    String strMsg = String.format(
                            "Received %d, To Receive %d",
                            o.m_iBytesSent, o.m_iBytesLeft);
                    rpcbase.JavaOutputMsg(strMsg);
                }
                int iSizeRecv = o.m_iBytesLeft;
                if(iSizeRecv > iSizeLimit)
                    iSizeRecv = iSizeLimit;
                else if(iSizeRecv == 0 ) {
                    onTransferDone(hChannel, 0 );
                    if(m_oInst.isServer())
                    {
                        IRpcService.IDeferredCall odc = new IRpcService.IDeferredCall() {
                            @Override
                            public int getArgCount() {
                                return 2;
                            }

                            @Override
                            public Class<?>[] getArgTypes() {
                                return new Class[]{
                                    long.class,
                                    byte[].class
                                };
                            }

                            @Override
                            public void call(Object[] oParams) {
                                long hChannel = (Long)oParams[0];
                                byte[] buf = (byte[]) oParams[1];
                                sendToken(hChannel, buf);
                            }
                        };
                        ret = m_oInst.deferCall(
                                odc, new Object[]{hChannel,
                                        "over".getBytes(StandardCharsets.UTF_8)});
                    }
                    break;
                }
                JRetVal jret = m_oInst.readStreamAsync(
                        hChannel, iSizeRecv );
                if(jret.ERROR()) {
                    ret = jret.getError();
                    break;
                }
                if(jret.isPending()) {
                    ret = RC.STATUS_PENDING;
                    break;
                }
                buf = (byte[]) jret.getAt(0);
            }
        }catch( IOException e){
            ret = -RC.EIO;
        }
        if(RC.ERROR(ret))
        {
            onTransferDone(hChannel,ret);
        }
        return ret;
    }

    synchronized public void onWriteStreamComplete(
            long hChannel, byte[] buf)
    {
        if(hChannel == RC.INVALID_HANDLE || buf == null)
            return;
        TransFileContext o = m_mapChanToCtx.get(hChannel);
        if(o == null)
            return;
        o.m_iBytesLeft -= buf.length;
        o.m_iBytesSent += buf.length;
        String strMsg = String.format(
                "Sent %d, To send %d",
                o.m_iBytesSent, o.m_iBytesLeft);
        rpcbase.JavaOutputMsg(strMsg);
        readFileAndSend(hChannel);
    }

    synchronized public void onReadStreamComplete(
            long hChannel, byte[]buf ) {
        if (hChannel == RC.INVALID_HANDLE || buf == null)
            return;
        TransFileContext o = m_mapChanToCtx.get(hChannel);
        if(o == null)
            return;
        writeFileAndRecv(hChannel,buf);
    }

    synchronized void setError(long hChannel, int iError)
    {
        if(hChannel == RC.INVALID_HANDLE)
            return;
        TransFileContext o = m_mapChanToCtx.get(hChannel);
        if(o == null)
            return;
        o.setError(iError);
    }

    synchronized int getError(long hChannel)
    {
        if(hChannel == RC.INVALID_HANDLE)
            return -RC.EINVAL;
        TransFileContext o = m_mapChanToCtx.get(hChannel);
        if(o == null)
            return -RC.EFAULT;
        return o.getError();
    }
}

File: examples/python/sftest/TransferContext.py
from rpcf.proxy import *
import threading as tr
import TransFileContext

class TransferContext :
    def __init__(self, oInst : PyRpcServices ) :
        self.m_oInst = oInst
        self.m_mapChanToCtx = dict()

    def AddContext(self, hChannel : int, o : TransFileContext )->int:
        if hChannel == ErrorCode.INVALID_HANDLE :
            return -errno.EINVAL
        old = None
        if hChannel in self.m_mapChanToCtx:
            old = self.m_mapChanToCtx[ hChannel ]

        if old is not None and not old.IsDone():
            return -errno.EEXIST

        self.m_mapChanToCtx[hChannel] = o
        return ErrorCode.STATUS_SUCCESS

    def RemoveContext( self, hChannel : int):
        self.m_mapChanToCtx.pop(hChannel)

    def OnTransferDone( self, hChannel : int, iRet : int) :
        if not hChannel in self.m_mapChanToCtx:
            return
        o = self.m_mapChanToCtx[hChannel]
        o.CleanUp()
        self.RemoveContext(hChannel)

    def ReadFileAndSend( self, hChannel : int)->int:
        ret = 0
        if not hChannel in self.m_mapChanToCtx:
            return -errno.ENOENT
        iSizeLimit = 2 * 1024 * 1024
        o = self.m_mapChanToCtx[hChannel]
        while True:
            iSize = o.m_iBytesLeft
            if iSize > iSizeLimit:
                iSize = iSizeLimit
            elif iSize == 0 :
                self.OnTransferDone(hChannel, 0)
                break
            pBuf = o.m_fp.read(iSize)
            ret = self.m_oInst.WriteStreamAsync2(
                hChannel, pBuf)
            if ret == ErrorCode.STATUS_PENDING:
                break
            if ret < 0:
                self.OnTransferDone(hChannel, ret)
                break
            o.m_iBytesLeft -= iSize
            o.m_iBytesSent += iSize

        return ret

    def SendToken( self, hChannel : int, buf : bytearray) -> int :
        return self.m_oInst.WriteStreamNoWait(hChannel,buf)

    def WriteFileAndRecv( self, hChannel:int, buf : bytearray ) -> int :
        ret = 0
        if hChannel == ErrorCode.INVALID_HANDLE:
            return -errno.EINVAL
        iSizeLimit = 2 * 1024 * 1024
        if not hChannel in self.m_mapChanToCtx:
            return -errno.ENOENT
        o = self.m_mapChanToCtx[hChannel]
        while True:
            iSize = len(buf)
            if o.m_iBytesLeft < iSize:
                ret = -errno.ERANGE
                break
            if iSize > 0:
                o.m_fp.write(buf)
                o.m_iBytesLeft -= iSize
                o.m_iBytesSent += iSize
                OutputMsg( "Received %d, To receive %d" %
                    ( o.m_iBytesSent, o.m_iBytesLeft ) )

            iSizeRecv = o.m_iBytesLeft
            if iSizeRecv > iSizeLimit:
                iSizeRecv = iSizeLimit
            elif iSizeRecv == 0:
                self.OnTransferDone(hChannel, 0)
                if self.m_oInst.IsServer() :
                    print("send over token")
                    self.m_oInst.DeferCall(
                        self.SendToken, hChannel, "over".encode())
                break

            pret = self.m_oInst.ReadStreamAsync2(
                hChannel, iSizeRecv )
            ret = pret[ 0 ]
            if ret < 0 or ret == ErrorCode.STATUS_PENDING:
                break
            buf = pret[1]
        
        if ret < 0:
            self.OnTransferDone(hChannel, ret)
        
        return ret

    def WriteStmCallback(self, hChannel : int, buf : bytearray) :
        if hChannel == ErrorCode.INVALID_HANDLE or buf is None:
            return
        if not hChannel in self.m_mapChanToCtx:
            return
        o = self.m_mapChanToCtx[ hChannel]
        o.m_iBytesLeft -= len(buf)
        o.m_iBytesSent += len(buf)
        OutputMsg( "Sent %d, To send %d" %
            (o.m_iBytesSent, o.m_iBytesLeft))
        self.ReadFileAndSend(hChannel)

    def ReadStmCallback(self, hChannel : int, buf : bytearray) :
        if hChannel == ErrorCode.INVALID_HANDLE or buf is None:
            return
        self.WriteFileAndRecv(hChannel, buf)

    def SetError( self, hChannel : int, iError : int) :
        if hChannel == ErrorCode.INVALID_HANDLE :
            return
        if not hChannel in self.m_mapChanToCtx:
            return
        o = self.m_mapChanToCtx[ hChannel ]
        o.SetError( iError )

    def GetError( self, hChannel : int ) -> int:
        if hChannel == ErrorCode.INVALID_HANDLE :
            return
        if not hChannel in self.m_mapChanToCtx:
            return -errno.ENOENT
        o = self.m_mapChanToCtx[ hChannel ]
        return o.GetError()

class TransferContextCli( TransferContext ):
    def __init__(self, oInst : PyRpcServices) :
        super().__init__(oInst)
        self.m_sem = tr.Semaphore(0)

    def OnTransferDone(self, hChannel: int, iRet: int):
        if hChannel == ErrorCode.INVALID_HANDLE:
            return
        if not hChannel in self.m_mapChanToCtx:
            return
        o = self.m_mapChanToCtx[ hChannel]
        if o.IsDone() :
            return
        o.SetError( iRet )
        o.CleanUp()
        self.NotifyComplete()

    def NotifyComplete( self ) :
        self.m_sem.release()

    def WaitForComplete( self ) :
        self.m_sem.acquire()
        
